### PR TITLE
pkg/{ipcache,policy}: use auto-generated code for DeepEqual

### PR DESCRIPTION
**Pull request by aanm**
To prevent missing fields from being DeepEqualed, we should use the
auto-generated functions as they will be able to detect and warn a
developer if new fields are added to a structure that is being
DeepEqualed.

Signed-off-by: André Martins <andre@cilium.io>

### CODE REVIEW

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

File: pkg/policy/api/utils.go [hunk 1]
```diff
@@ -84,30 +52,28 @@ func (d *PortRuleDNS) Exists(rules L7Rules) bool {
...
-		return false
-	}
-	for k, v := range *h {
-		if v2, ok := o[k]; !ok || v2 != v {
-			return false
-		}
+// DeepEqual returns true if both L7 rules are equal
+func (h *PortRuleL7) DeepEqual(o *PortRuleL7) bool {
+	if h == nil {
+		return o == nil || *o == nil
```
Comment (jrajahalme) [inline pkg/policy/api/utils.go:74]: Haven't checked, but do we need `return o == nil || *o == nil` in all `DeepEqual()` functions when `*o` is a map or a slice? Would `return o == nil || len(*o) == 0` have the same effect?

File: pkg/policy/l4.go [hunk 2]
```diff
@@ -105,15 +104,23 @@ func __canSkip(currentRule *PerSelectorPolicy, wildcardRule *PerSelectorPolicy)
...
 type TLSContext struct {
 	TrustedCA        string `json:"trustedCA,omitempty"`
 	CertificateChain string `json:"certificateChain,omitempty"`
 	PrivateKey       string `json:"privateKey,omitempty"`
 }
 
-// Equal returns true if 'a' and 'b' have the same contents.
-func (a *TLSContext) Equal(b *TLSContext) bool {
-	return a == nil && b == nil || a != nil && b != nil && *a == *b
+// DeepEqual is an autogenerated deepequal function, deeply comparing the
```
Comment (jrajahalme) [inline pkg/policy/l4.go]: Comment is a bit off, it's `deepEqual()` that is the autogenerated function.

File: pkg/policy/l4.go [hunk 3]
```diff
@@ -165,13 +178,17 @@ type PerSelectorPolicy struct {
...
-// Equal returns true if 'a' and 'b' represent the same L7 Rules
-func (a *PerSelectorPolicy) Equal(b *PerSelectorPolicy) bool {
-	return a == nil && b == nil || a != nil && b != nil &&
-		a.TerminatingTLS.Equal(b.TerminatingTLS) &&
-		a.OriginatingTLS.Equal(b.OriginatingTLS) &&
-		a.IsDeny == b.IsDeny &&
-		reflect.DeepEqual(a.L7Rules, b.L7Rules)
+// Equal is used in tests for the check.Checker.
+func (in *PerSelectorPolicy) Equal(o *PerSelectorPolicy) bool {
+	return in.DeepEqual(o)
```
Comment (jrajahalme) [inline pkg/policy/l4.go:183]: This is not handling the case `in == nil` like the old `Equal()` did. If this would, then this becomes the same as `DeepEqual()` below.

File: pkg/policy/l4.go [hunk 4]
```diff
@@ -153,10 +162,14 @@ type PerSelectorPolicy struct {
...
 	// Computed after rule merging is complete!
+	//
+	// +deepequal-gen=false
 	EnvoyHTTPRules *cilium.HttpNetworkPolicyRules `json:"-"`
 
 	// CanShortCircuit is true if all 'EnvoyHTTPRules' may be
 	// short-circuited by other matches.
+	//
+	// +deepequal-gen=false
 	CanShortCircuit bool `json:"-"`
```
Comment (jrajahalme) [inline pkg/policy/l4.go:173]: I feel like these fields *should* be included in the new `DeepEqual()`, but kept excluded from `Equal()`. One way of achieving this would be to keep them excluded from autogenerated `deepEqual()` and manually compare them in `DeepEqual()`?

File: pkg/policy/portmap.go [hunk 5]
```diff
@@ -44,14 +46,18 @@ type NamedPortMap map[string]PortProto
 // PortProtoSet is a set of unique PortProto values.
 type PortProtoSet map[PortProto]struct{}
 
-// Equal returns true if the PortProtoSets are equal.
-func (pps PortProtoSet) Equal(other PortProtoSet) bool {
-	if len(pps) != len(other) {
+// DeepEqual returns true if the PortProtoSets are equal.
+func (pps PortProtoSet) DeepEqual(other *PortProtoSet) bool {
```
Comment (jrajahalme) [inline pkg/policy/portmap.go:50]: Why this change when there is nothing that is autogenerated here? You want the call sites be written as if `DeepEqual()` used autogenerated code?

File: pkg/policy/selectorcache.go [hunk 6]
```diff
@@ -256,12 +256,6 @@ type selectorManager struct {
 	cachedSelections map[identity.NumericIdentity]struct{}
 }
 
-// Equal is used by checker.Equals, and only considers the identity of the selector,
-// ignoring the internal state!
-func (s *selectorManager) Equal(b *selectorManager) bool {
-	return s.key == b.key
-}
-
```
Comment (jrajahalme) [inline pkg/policy/selectorcache.go:264]: Was this unused?

**Review by jrajahalme — APPROVED** [review]
Only minor nits :-)

File: pkg/policy/api/utils.go (hunk 1 above)
Comment (aanm) [reply pkg/policy/api/utils.go:74]: >Haven't checked, but do we need return o == nil || *o == nil in all DeepEqual() functions when *o is a map or a slice? 

yes, `o` is the pointer not the map / slice. Either `o` or `*o` can be nil.

> Would return o == nil || len(*o) == 0 have the same effect?

it would have the same effect, would you consider that DeepEqual? We would be comparing a nil map `h` with a non-nil map `o`, not sure if they would be considered them DeepEqual? :thinking:

File: pkg/policy/l4.go (hunk 4 above)
Comment (aanm) [reply pkg/policy/l4.go:173]: it's up to us to make it consistent. If we have different implementations of DeepEqual and Equal, why should one be ignoring some fields but not the other? i.e., the purpose of this PR was to avoid developers missing the addition of new fields into these structures.

File: pkg/policy/l4.go (hunk 3 above)
Comment (aanm) [reply pkg/policy/l4.go:183]: I don't understand the comment, can you clarify it? This `Equal` needs to be present because of `check.Checker`.

File: pkg/policy/portmap.go (hunk 5 above)
Comment (aanm) [reply pkg/policy/portmap.go:50]: Yes. The autogenerated structures that use this field will call DeepEqual for this structure so I end up renaming this function as well.

File: pkg/policy/selectorcache.go (hunk 6 above)
Comment (aanm) [reply pkg/policy/selectorcache.go:264]: Yes, I didn't find any usage of it and tests haven't failed.

**Comment by aanm** [comment]
test-me-please

File: pkg/policy/api/utils.go (hunk 1 above)
Comment (jrajahalme) [reply pkg/policy/api/utils.go:74]: I would consider two maps equal if they have the same elements, regardless the actual representation of an empty map, especially if that does not make a difference for the use of the map. But maybe for testing purposes we'd want DeepEqual to actually verify that the data structures have the same structure, and not only that they are functionally equivalent? On that note the code above is stating a `nil` pointer to a map `h` is DeepEqual to a non-nil pointer to a nil map `o`.

File: pkg/policy/l4.go (hunk 3 above)
Comment (jrajahalme) [reply pkg/policy/l4.go:183]: The old implementation allowed `in` to be `nil`, the new does not.

File: pkg/policy/l4.go (hunk 4 above)
Comment (jrajahalme) [reply pkg/policy/l4.go:173]: `EnvoyHTTPRules` is included in this structure because it needs to be computed while the context for secret manager is available. It has been excluded from `Equal` to allow unit tests to behave as if it did not exist. Might be better to amend the unit tests to include the Envoy translation of the rules, but on the other hand it would complicate the policy package unit tests with envoy package implementation details.

File: pkg/policy/l4.go (hunk 3 above)
Comment (jrajahalme) [reply pkg/policy/l4.go:183]: Not sure it is needed though, if you did not see any panics due to `nil` pointer dereference in `in.DeepEqual()`?

File: pkg/policy/api/utils.go (hunk 1 above)
Comment (aanm) [reply pkg/policy/api/utils.go:74]: @jrajahalme I took a look at go docs for the `reflect.DeepEqual`:
```
// Map values are deeply equal when all of the following are true:
// they are both nil or both non-nil, they have the same length,
// and either they are the same map object or their corresponding keys
// (matched using Go equality) map to deeply equal values.

// Slice values are deeply equal when all of the following are true:
// they are both nil or both non-nil, they have the same length,
// and either they point to the same initial entry of the same underlying array
// (that is, &x[0] == &y[0]) or their corresponding elements (up to length) are deeply equal.
// Note that a non-nil empty slice and a nil slice (for example, []byte{} and []byte(nil))
// are not deeply equal.
```

so I'll just perform the same behavior as the `reflect.DeepEqual`
